authority/tls.go

```diff
@@ -135,92 +136,17 @@ func (a *Authority) Sign(csr *x509.CertificateRequest, signOpts provisioner.Opti
-		MaxPathLen:                  oldCert.MaxPathLen,
-		MaxPathLenZero:              oldCert.MaxPathLenZero,
-		OCSPServer:                  oldCert.OCSPServer,
-		IssuingCertificateURL:       oldCert.IssuingCertificateURL,
-		PermittedDNSDomainsCritical: oldCert.PermittedDNSDomainsCritical,
-		PermittedEmailAddresses:     oldCert.PermittedEmailAddresses,
-		DNSNames:                    oldCert.DNSNames,
-		EmailAddresses:              oldCert.EmailAddresses,
-		IPAddresses:                 oldCert.IPAddresses,
-		URIs:                        oldCert.URIs,
-		PermittedDNSDomains:         oldCert.PermittedDNSDomains,
-		ExcludedDNSDomains:          oldCert.ExcludedDNSDomains,
-		PermittedIPRanges:           oldCert.PermittedIPRanges,
-		ExcludedIPRanges:            oldCert.ExcludedIPRanges,
-		ExcludedEmailAddresses:      oldCert.ExcludedEmailAddresses,
-		PermittedURIDomains:         oldCert.PermittedURIDomains,
-		ExcludedURIDomains:          oldCert.ExcludedURIDomains,
-		CRLDistributionPoints:       oldCert.CRLDistributionPoints,
-		PolicyIdentifiers:           oldCert.PolicyIdentifiers,
-	}
-
-	// Copy all extensions except for Authority Key Identifier. This one might
-	// be different if we rotate the intermediate certificate and it will cause
-	// a TLS bad certificate error.
-	for _, ext := range oldCert.Extensions {
-		if !ext.Id.Equal(oidAuthorityKeyIdentifier) {
-			newCert.ExtraExtensions = append(newCert.ExtraExtensions, ext)
-		}
-	}
-
-	leaf, err := x509util.NewLeafProfileWithTemplate(newCert, a.x509Issuer, a.x509Signer)
-	if err != nil {
-		return nil, errs.Wrap(http.StatusInternalServerError, err, "authority.Renew", opts...)
-	}
-	crtBytes, err := leaf.CreateCertificate()
-	if err != nil {
-		return nil, errs.Wrap(http.StatusInternalServerError, err,
-			"authority.Renew; error renewing certificate from existing server certificate", opts...)
-	}
-
-	serverCert, err := x509.ParseCertificate(crtBytes)
-	if err != nil {
-		return nil, errs.Wrap(http.StatusInternalServerError, err,
-			"authority.Renew; error parsing new server certificate", opts...)
-		}
-
-	if err = a.db.StoreCertificate(serverCert); err != nil {
-		if err != db.ErrNotImplemented {
-			return nil, errs.Wrap(http.StatusInternalServerError, err, "authority.Renew; error storing certificate in db", opts...)
-		}
-	}
-
-	return []*x509.Certificate{serverCert, a.x509Issuer}, nil
+	return a.RenewOrRekey(oldCert, oldCert.PublicKey)
 }
 
-// Rekey is similar to renew except that the certificate will be renewed with new key from csr argument.
-func (a *Authority) Rekey(oldCert *x509.Certificate, csr *x509.CertificateRequest) ([]*x509.Certificate, error) {
+// Rekey is similar to renew except that the certificate will be renewed with new key.
+// Function does rekeying if a new public key is passed as pk else if same key is passed, certificate will be just renewed.
+func (a *Authority) RenewOrRekey(oldCert *x509.Certificate, pk crypto.PublicKey) ([]*x509.Certificate, error) {
```

Comment:
Is this function fine? I am calling this from Renew. Not sure if RenewOrRekey is the right name, couldn't think of anything else. And for the test case, shall i modify the existing TestAuthority_Renew to test RenewOrRekey (i.e TestAuthority_RenewOrRekey)?

Reply:
I think the function should be called just `Rekey`. It's cleaner. But we can let @maraino weigh in as well.

Reply:
Just `Rekey`. and calling just Rekey as you did in Renew.